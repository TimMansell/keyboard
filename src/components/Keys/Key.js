import React from 'react';
import PropTypes from 'prop-types';
import Tone from 'tone';
import './key.scss';

function Key({ type }) {
  const synth = new Tone.Synth({
    oscillator: {
      partials: [3, 2, 1],
      type: 'custom',
      frequency: 'C#4',
      volume: -12,
    },
    envelope: {
      attack: 0.11,
      decay: 0.21,
      sustain: 1,
      release: 1.71,
      attackCurve: 'exponential',
      decayCurve: 'exponential',
      releaseCurve: 'exponential',
    },
  }).toMaster();

  const handleClick = (key) => {
    synth.triggerAttackRelease(key, '8n');
  };

  return (
    <div className="key">
      {type.map((key, index) => (
        <div
          className="key__btn"
          key={index}
          onClick={() => handleClick(key)}
        ></div>
      ))}
    </div>
  );
}

Key.propTypes = {
  type: PropTypes.array.isRequired,
};

export default Key;
