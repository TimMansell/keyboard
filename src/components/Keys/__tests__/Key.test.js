import React from 'react';
import { render } from '@testing-library/react';

jest.mock('../../../service/audio');
jest.mock('tone');
import playKey from '../../../service/audio';
import Tone from 'tone';
import Key from '../Key';

const oneNote = ['E3'];
const twoNotes = ['C3', 'C#3'];

test('one white key snapshot', () => {
  const { asFragment } = render(<Key type={oneNote} />);

  expect(asFragment()).toMatchSnapshot();
});

test('one white and one black key snapshot', () => {
  const { asFragment } = render(<Key type={twoNotes} />);

  expect(asFragment()).toMatchSnapshot();
});

test('render one key', () => {
  const { getByTestId, getAllByTestId } = render(<Key type={oneNote} />);

  expect(getByTestId('key-0')).toBeTruthy();
  expect(getAllByTestId(/key/).length).toBe(1);
});

test('render two keys', () => {
  const { getByTestId, getAllByTestId } = render(<Key type={twoNotes} />);

  expect(getByTestId('key-0')).toBeTruthy();
  expect(getByTestId('key-1')).toBeTruthy();
  expect(getAllByTestId(/key/).length).toBe(2);
});
