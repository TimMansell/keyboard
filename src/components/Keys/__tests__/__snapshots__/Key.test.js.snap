// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one white and one black key snapshot 1`] = `
<DocumentFragment>
  <div
    class="key"
  >
    <div
      class="key__btn"
      data-testid="key-0"
    />
    <div
      class="key__btn"
      data-testid="key-1"
    />
  </div>
</DocumentFragment>
`;

exports[`one white key snapshot 1`] = `
<DocumentFragment>
  <div
    class="key"
  >
    <div
      class="key__btn"
      data-testid="key-0"
    />
  </div>
</DocumentFragment>
`;
